fuse_core/include/fuse_core/loss.h

```diff
@@ -0,0 +1,307 @@
+   * generated loss functions (i.e. the Ceres objects will be destroyed before the Loss Function objects). This
+   * guarantee may allow optimizations for the creation of the loss function objects.
+   *
+   * @return A base pointer to an instance of a derived ceres::LostFunction.
+   */
+  ceres::LossFunction* lossFunction() const
+  {
+    return loss_function_;
+  }
+
+  /**
+   * @brief Perform a deep copy of the Loss and return a unique pointer to the copy
+   *
+   * This can/should be implemented as follows in all derived classes:
+   * @code{.cpp}
+   * return Derived::make_unique(*this);
+   * @endcode
+   *
+   * @return A unique pointer to a new instance of the most-derived Loss
+   */
+  virtual Loss::UniquePtr clone() const = 0;
+
+  /**
+   * @brief Serialize this Loss into the provided binary archive
+   *
+   * This can/should be implemented as follows in all derived classes:
+   * @code{.cpp}
+   * archive << *this;
+   * @endcode
+   *
+   * @param[out] archive - The archive to serialize this loss function into
+   */
+  virtual void serialize(fuse_core::BinaryOutputArchive& /* archive */) const = 0;
+
+  /**
+   * @brief Serialize this Loss into the provided text archive
+   *
+   * This can/should be implemented as follows in all derived classes:
+   * @code{.cpp}
+   * archive << *this;
+   * @endcode
+   *
+   * @param[out] archive - The archive to serialize this loss function into
+   */
+  virtual void serialize(fuse_core::TextOutputArchive& /* archive */) const = 0;
+
+  /**
+   * @brief Deserialize data from the provided binary archive into this Loss
+   *
+   * This can/should be implemented as follows in all derived classes:
+   * @code{.cpp}
+   * archive >> *this;
+   * @endcode
+   *
+   * @param[in] archive - The archive holding serialized Loss data
+   */
+  virtual void deserialize(fuse_core::BinaryInputArchive& /* archive */) = 0;
+
+  /**
+   * @brief Deserialize data from the provided text archive into this Loss
+   *
+   * This can/should be implemented as follows in all derived classes:
+   * @code{.cpp}
+   * archive >> *this;
+   * @endcode
+   *
+   * @param[in] archive - The archive holding serialized Loss data
+   */
+  virtual void deserialize(fuse_core::TextInputArchive& /* archive */) = 0;
+
+private:
+  ceres::LossFunction* loss_function_{ nullptr };  //!< The ceres::LossFunction
+
+  // Allow Boost Serialization access to private methods
+  friend class boost::serialization::access;
+
+  /**
+   * @brief The Boost Serialize method that serializes all of the data members in to/out of the archive
+   *
+   * @param[in/out] archive - The archive object that holds the serialized class members
+   * @param[in] version - The version of the archive being read/written. Generally unused.
+   */
+  template<class Archive>
+  void serialize(Archive& /* archive */, const unsigned int /* version */)
```

Comment:
True, I just gave that a try, but I relaized that in order to do that I need to provide Boost serialization support to `ceres::LossFunction`s, which I don't think it's possible, or at least it's not trivial.

I've removed the `loss_function_` attribute from the `Loss` base class and now it's the children responsibility to return a `ceres::LossFunction*` from the `lossFunction()` method, that now is pure virtual.

The ownership of the `ceres::LossFunction` has also been changed to `TAKE_OWNERSHIP` (Ceres default), although it could be changed to `DO_NOT_TAKE_OWNERSHIP`.

## Thoughts on supporting `DO_NOT_TAKE_OWNERSHIP`

First of all, this  could be done in the future, in another PR.

I believe the hardest part to do so would be the Boost serialization:
* I think it'd required a split `save` and `load`, so we can construct the `ceres::LossFunction` in the `load` method only.
* It'd be constructed based on the other members in the children classes (typically the `a` argument, which would be the only thing de/serialized).
* The children would have an `std::shared_ptr<ceres::LossFunction> loss_function_` (or `std::unique_ptr`) member and they'd return a RAW pointer to it with `.get()` in `lossFunction()`.